### ISSUE TITLE
Music stops after listening for an extended period of time

### ISSUE THREAD

**Issue by ahmaxed** [freeCodeCamp/freeCodeCamp#41074]
Although music stops after after listening for an extended period of time, it is restored after switching between the streams. 
Have you ran across such problem? @huyenltnguyen

**Comment by huyenltnguyen** [comment]
I'm not sure if this piece of information is useful, but I was trying to compare the playing time between the Azuracast player and ours. My assumption was that there was an issue with WebSocket connection, and I expected the Azuracast player, which uses the [standard method](https://www.azuracast.com/developers/nowplaying.html#standard-now-playing-api), to last longer.

I opened them in two tabs, kept them both playing but muted one and occasionally switched the mute over (to ensure they are still playing). When the music stopped on one tab, I found that the other stopped as well, which is... interesting. Though I still have no clue on what the cause is.

**Comment by ahmadabdolsaheb** [comment]
Were both on the background? might be the browser deciding to close background tab tasks. We might want to look into background tab optimization if that is the case.

**Comment by huyenltnguyen** [comment]
Yeah they were both running on the background.

But according to [MDN](https://developer.mozilla.org/en-US/docs/Web/API/Page_Visibility_API#policies_in_place_to_aid_background_page_performance) we shouldn't worry about how browsers treat them. 
> Some processes are exempt from this throttling behavior. In these cases, you can use the Page Visibility API to reduce the tabs' performance impact while they're hidden.
> - Tabs which are playing audio are considered foreground and aren’t throttled.
> - Tabs running code that's using real-time network connections (WebSockets and WebRTC) go unthrottled in order to avoid closing these connections timing out and getting unexpectedly closed.
> - IndexedDB processes are also left unthrottled in order to avoid timeouts.
closing these connections timing out and getting unexpectedly closed.

I wonder if this is an AzuraCast issue, although I can't seem to find anything similar looking through their issue list.

**Comment by ahmadabdolsaheb** [comment]
@huyenltnguyen, have you been experiencing this issue recently?
It seems like chrome has introduced some new optimizations that throttles the performance of background tabs.

**Comment by huyenltnguyen** [comment]
I'm on Firefox most of the time and I have been getting the same experience, in that the music stops after around 30 mins of playing. Had the experienced been improved for you until that Chrome update?

**Comment by ahmadabdolsaheb** [comment]
It was working well for me for 4- 5 months. I went ahead and updated the azuracast azurarelay and all their vms, and will be remove and read the components to isolate the issue locally.

**Comment by huyenltnguyen** [comment]
Thanks, Ahmad! Let me know if you find anything suspicious. 

On my end, I recently found that the issue sometimes can self-recover after 2 or 3 hours of stoppage. Though I wasn't consciously aware or keeping track of this, so I don't have any concrete conclusion yet.

**Comment by huyenltnguyen** [comment]
I saw some errors in my console, not sure if they are thrown at the time the music stops, but I did a quick search and found a lot of search results for this. It appears to be a WebSocket-on-Firefox issue. This could at least explain my experience with CodeRadio since I have been using it mostly on Firefox, but I'll try with Chrome to see how the app performs there.

<img width="1680" alt="Screen Shot 2021-07-29 at 13 45 24" src="https://user-images.githubusercontent.com/25715018/127444745-6336c9ff-0083-42f6-b15e-150e9fe19b32.png">

**Comment by iepsen** [comment]
What browser and operating system are you using? I can't repro the issue using Chrome Version 95.0.4638.69 (Official Build) (x86_64) with macOS Monterey Version 12.0.1 (21A559). It's playing from around 2 hours without any interruption on main branch locally within a slow network preset.

**Comment by ahmadabdolsaheb** [comment]
@iepsen, I have been experienced this more often when using a lower end device under load.
Also, here is a related issue: https://github.com/AzuraCast/AzuraCast/issues/1420

**Comment by ahmadabdolsaheb** [comment]
I am playing coderadio on a simple html js on codepen and still getting the errors. it seems to be an issue with browsers throttling html5 audio elements in background tabs. We need a way to keep the tab "active" and catch the connection error and replay accordingly.

Here is the codepen: https://codepen.io/abdolsa/pen/dyVdgNM

**Comment by iepsen** [comment]
This is what I see on console:
> Uncaught (in promise) DOMException: play() failed because the user didn't interact with the document first

Is it the same you are seeing?

**Comment by ahmadabdolsaheb** [comment]
If you are referring to the codepen, some browsers error out because the audio plays automatically. I added a button so it should not show you that error.

This is the error:
`GET https://coderadio-relay-blr.freecodecamp.org/radio/8010/radio.mp3 net::ERR_CONNECTION_CLOSED 200`

Here is a summary of the issue: https://stackoverflow.com/questions/70532842/how-to-avoid-background-tab-throttling-for-an-html5-radio-app

**Comment by Sboonny** [comment]
I had an issue similar to this. I lost the connection, paused the radio, then tried to resume the radio but couldn't.

When I lost my connection got this

 >NchanSubscriber.js:664      
       WebSocket connection to 'wss://coderadio-admin.freecodecamp.org/api/live/nowplaying/coderadio?last_event_id=1655700138%3A0' failed: 

When I paused with connection lost, then unpaused got this

>coderadio.freecodecamp.org/:1       
       Uncaught (in promise) DOMException: The element has no supported sources.
Promise.then (async)
value @ App.js:243
value @ App.js:288
e.handleOnTouchEnd @ PlayPauseButton.js:15
s @ react-dom.production.min.js:14
p @ react-dom.production.min.js:14
(anonymous) @ react-dom.production.min.js:14
g @ react-dom.production.min.js:15
st @ react-dom.production.min.js:52
at @ react-dom.production.min.js:51
ut @ react-dom.production.min.js:52
ht @ react-dom.production.min.js:56
D @ react-dom.production.min.js:287
q @ react-dom.production.min.js:19
Zt @ react-dom.production.min.js:70
Jt @ react-dom.production.min.js:69
t.unstable_runWithPriority @ scheduler.production.min.js:19
$i @ react-dom.production.min.js:122
M @ react-dom.production.min.js:287
Xt @ react-dom.production.min.js:68

![image](https://user-images.githubusercontent.com/88248797/174529610-00cecfbc-8374-4cb3-a8c5-4197ccb13f10.png)

**Comment by michalsieron** [comment]
I know it's been a long time since this issue was touched, but I have some observations that weren't mentioned in this thread yet (actually something very similar was, but it was still in the browser: https://github.com/freeCodeCamp/freeCodeCamp/issues/41074#issuecomment-737273843).

I am listening to the radio with mpv like so:
```
mpv https://coderadio-admin.freecodecamp.org/radio/8010/radio.mp3
```

and I also experience this issue that stream sometimes stops.

Now when music stops, I usually get `Audio device underrun detected.` error from mpv, and I see that the cache reported by the mpv is empty. I tried many times to find if this is a pipewire configuration issue, but this is the only audio issue I get, so it seems it's a stream issue.
Usually after a second or two, the cache fills up again and music is back, but sometimes it just stops and that's it.

Sometimes I also get `[ffmpeg/audio] invalid new backstep -1` error. This indicates it's a deeper issue, because the codec itself is complaining (not sure, not an audio engineer).

After I've seen such error multiple times, I decided to try without mpv, and I was listening to the radio like this (before this I also tried piping to the mpv, but with exact same results):
```
curl https://coderadio-admin.freecodecamp.org/radio/8010/radio.mp3 | ffplay -nodisp -autoexit -i -
```

Running the stream this way seemed more stable at first, but at some point I got a familiar error: `[mp3float @ 0x7efcbc00ccc0] invalid new backstep -1`.

The stream did not always end with the above error. In fact, it was rather rare to get such error. Most of the time I was getting: `HTTP/2 stream 1 was not closed cleanly: INTERNAL_ERROR (err 2)`.

Do with this what you want, it is not consistent to say how to reproduce exactly, but I just thought this may help.

btw, thanks for awesome radio :heart:

**Comment by michalsieron** [comment]
First, reply to your question @huyenltnguyen: I am using EndeavourOS (Linux distro, Arch based) and I do not use a browser to listen to the radio :slightly_smiling_face: 

---

I wanted to debug it further myself before replying, but I won't probably have time to do such.
So here are some things I found a few days after my original reply.

\<tldr>

I did three sets of tests:
- laptop A with VPN and proxy, 8 `curl`s redirecting to `/dev/null`, all died after some time
- laptop B (without VPN or proxy) with the same behavior as laptop A with both `curl` and `fetch`
- again laptop B with different behavior, with `curl`, `curl --http1.1`, `wget` and `aria2c`

For `curl --http1.1` failures and one `aria2c` failure, error was with the TLS.

\</tldr>

---

When I was writing my first reply, I noticed that if I get `HTTP/2 stream 1 was not closed cleanly: INTERNAL_ERROR (err 2)` error, then it is not really a codec issue. It is not even an issue with the data being sent and being incorrectly treated.

So what I did, was to run eight `curl https://coderadio-admin.freecodecamp.org/radio/8010/radio.mp3 > /dev/null` in parallel and see what happens. They all failed with the same `HTTP/2` error as before after: 36m4s, 40m43s, 41m49s, 1h30m43s, 1h43m59s, 1h53m31s, 2h30m32s, 2h32m37s.

_Now let's take a small break and let me say that everything I've been testing up till now (so this part and previous reply) was on a laptop connected through a VPN, proxy and who knows what else. So I wouldn't be surprised if that was the culprit._

---

_Tests described below were performed on a different machine with technically same OS. It is not using any VPN or proxy._

So previous findings would suggest it is a pure network/server issue.

I thought there is a tiny-teensy possibility it is a curl issue, but it is so battle tested I didn't think it was possible. And sure enough, after performing the test with [fetch](https://ftp.netbsd.org/pub/pkgsrc/current/pkgsrc/net/fetch/index.html) from NetBSD (which doesn't use libcurl) I confirmed it was not a curl issue.
Eight `fetch` jobs failed after near 2 hours of playing to `/dev/null`. At the same time I was performing another `curl` test and all eight jobs finished around 1h40m with familiar `HTTP/2` error.

_So the issue was reproduced, and it is not related to the VPN and proxy on the previous machine?
Well, it is not so simple._

I decided to run another batch of tests. Now with `curl`, `curl --http1.1`, `wget` and `aria2c`.
Real tragedy is that, I had to restart my machine and I don't have logs for them :angry:.

Anyway, I still remembered some stuff when looking at the console.
First things first, no job died before 2 hours. Not even 3 or 4 hours. All jobs that died did so after about 50 hours (I don't know if me starting them after 6 PM on Friday is related :thinking:).

Some of them were even still running.

For example, `wget` kept restarting the download when it encountered an issue (it had a couple of them, but didn't say anything what happened, only that it was restarting the download).

For some reason, I had two `aria2c` download jobs running. One was logging to a file (I do have it :tada:. It's ~2 GB).
The one logging didn't die, but now when grepping the log file, I see 3 occurrences of `[DownloadCommand.cc:234] errorCode=1 Got EOF from the server.` error and 1 occurrence of `[SocketCore.cc:886] errorCode=1 Failed to receive data, cause: Error in the pull function.` error. Despite those errors, `aria2c` simply restarted the download like `wget` and continued working.
The second `aria2c` job, which again I don't have the logs for :angry:, died with some TLS error. Now that's getting interesting. Would need a way to reproduce it, but this could take us somewhere.

As for `curl` jobs, they went similarly to previous tests, with the difference being time after they died.
Now `curl --http1.1` behaved differently. For once, it can't show us `HTTP/2` error if it is not using HTTP2 :smile:. What it did show was some TLS error.
And yet again, I don't have the error message as it was. But what I have is power of browsing history :laughing:
I found these two searches in my history `TLSv1.3 (OUT), TLS alert, close notify (256):` and `curl: (18) transfer closed with outstanding read data remaining`.

Sorry for this information dump, which due to not being really reproduced yet, may not be that valuable :bowing_man:

**Comment by jamesgeddes** [comment]
I am seeing this too. Currently using,

- Chrome Version 120.0.6099.216 (Official Build) (64-bit)
- Ubuntu 22.04.3 LTS

**Comment by Sembauke** [comment]
I am going to close this as won't fix. Please re-open if you think otherwise.

**Comment by raisedadead** [comment]
Please open a fresh issue -- if unreproducible, or hard to investigate and fix add a warning, faq on the page instead. Closing
